_posts/jared_baker/2018-12-15-Bluestore.md

```diff
@@ -0,0 +1,171 @@
+
+<figure>
+    <img id="filestore-vs-bluestore-2" src="{{site.urlimg}}jared_baker/bluestore/filestore-vs-bluestore-2.png" data-featherlight="#filestore-vs-bluestore-2" />
+</figure>
+
+
+## Prequisites
+Since Bluestore was first officially supported in Luminous we needed to upgrade from Ceph Jewel to Luminous. We use Ubuntu 16.04 and the official ceph packages for that distro so it was a straight forward upgrade of adding the repo and upgrading the packages on our mon's, radosgw's and OSD nodes.
+<figure>
+    <img id="luminous_logo" src="{{site.urlimg}}jared_baker/bluestore/luminous_logo-1-300x116.png" data-featherlight="#luminous_logo" />
+</figure>
+
+## Migration Approach
+Drain - Convert - Fill!
+
+There is no 'conversion' from Filestore to Bluestore without major data movement so our approach was to drain a storage node of its data, followed by destroying the OSD and then prepare the OSD's using Bluestore followed by the refilling of the data from the cluster. Rinse and repeat for each storage node in the cluster. Depending on your ceph architecture you may be able to cycle several storage nodes in parallel to reduce the overall migration time. We would migrate 2-3 storage nodes at a time, each in different racks (our failure domain) and tried to keep below a high water mark of 20% 'misplaced objects' during the drain and fills.
+
+In order for this to be successful you have to ensure that your cluster has enough free space to drain each node to account for the temporary loss of cluster capacity when a node has to drain all it's data to the rest of the cluster.
+
+### Draining
+To drain a node we would modify the osd crush weight of all of a storage nodes OSD's to 0. This will trigger an evacuation of data away from this storage node and distribute it to the rest of the cluster.
+
+Example, to drain OSD's 720 to 755 we would run:
+~~~bash
+for i in $(seq 720 755); do ceph osd crush reweight osd.$i 0; done
+~~~
+
+When the draining is complete you can check ceph's health (ceph -s) to confirm if it is done draining the data. You can also use 'ceph osd df tree' to see how much data the disks still have to transfer off. Our storage nodes took on average 24 hours per server to drain.
+
+### Converting the OSD to Bluestore
+Follow these steps to safely retire the OSD and bring it back as Bluestore.
+[Available in script form here](https://github.com/CancerCollaboratory/infrastructure/blob/master/migrate_bluestore.sh)
+
+~~~bash
+# Stop the OSD process
+systemctl stop ceph-osd@<osd-id>.service
+
+# Unmount the OSD
+umount /dev/<block-device>
+
+# Zap the disk
+ceph-disk zap <osd-id>
+
+# Mark the OSD as destroyed
+ceph osd destroy <osd-id> --yes-i-really-mean-it
+
+# Prepare the disk as Bluestore
+ceph-disk prepare --bluestore /dev/<block-device> --osd-id <osd-id>
+~~~
+
+Notice that when comparing the output of ceph osd tree before/after conversion that the class column may now indicate your osd type (ssd or hdd) which for us, was blank before.
```

Comment:
- Why was it blank before the conversion for our implementation?  Is this a useful detail to add in for the reader?